### DISCUSSION TITLE
Even when the batch size is 1, CUDA is out of memory. What should I do?

### DISCUSSION THREAD

**Discussion by zuowanbushiwo** [speechbrain/speechbrain#1849] [Q&A]
1. I want to train a sepformer model for speech separation, path: speechbrain/recipes/LibriMix/separation/hparams/sepformer-libri2mix.yaml
2. I have 4 RTX3090 graphics cards, each with 24G memory, as shown below:
![image](https://user-images.githubusercontent.com/3354428/219999955-b19b6662-7b9f-46b4-bf34-4ee0fede841a.png)
3. The training reported the following error:
`RuntimeError: CUDA out of memory. Tried to allocate 212.00 MiB (GPU 0; 23.70 GiB total capacity; 21.03 GiB already allocated; 96.81 MiB free; 22.07 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF`

**Comment by pplantinga** [comment]
How long is the audio you are using here? If it is very long, that may cause the OOM error, and you may want to segment the audio before passing to the separation model.

**Reply by neurence** [reply]
I have the same issue.
In the past I've solved this in one of two ways:  Either by pre-segmenting the audio files into smaller chunks, as you suggest,  or by writing my own dataloader to do this on the fly, and not to touch the dataset.

I'm at a loss how to implement the latter way (which I much prefer) as a Dynamic Item Pipeline function which yields succesive chunks of an audio file.  How could I do this? I do know that dataio.read_audio allows me to specify a start/end offset, but this doesn't really help.

**Reply by pplantinga** [reply]
Perhaps @Gastron or @popcornell can give a better answer, but my understanding is that each item in the prepared json manifest should correspond to a chunk of audio in this case. Your train.json could look something like:

```json
{ "utterance_id_1": {"path": "/path/to/audiofile.wav", "start": 0, "end": 17304}, "utterance_id_2": {"path": "/path/to/audiofile.wav", "start": 17305, "end": 34568}...}
```

**Reply by neurence** [reply]
Yes, I've considered that solution, which is certainly achievable.  Your reply answers my implicit question, which is whether the items yielded in a data pipeline must correspond one-to-one with the keys in the manifest.   The two issues I have with your solution is 1) the bloating of the dataset dict, which is kept in memory, and 2)  using read_audo will open and close files many times -- unless I implement a stateful version of it, which is not trivial, due to there being multiple worker threads.

**Reply by Gastron** [reply]
Yep, the DynamicItemDataset is designed to have a static entry for each training example - it doesn't handle dynamically created examples. You've thought of some smart hacks, but I think you will ultimately have a better experience simply working with some other data loading solution in this case.

In SpeechBrain we have some tooling for WebDataset, and I think the new PyTorch DataPipes too, work with iterators, which handle these sorts of operations nicely.

**Reply by neurence** [reply]
FYI I created this function to serve segments of files without continually opening and closing them.  The code is not yet robust.  It uses a manifest as in @pplantinga 's post above

```
    @sb.utils.data_pipeline.takes("file_path", "start_offset", "length")
    @sb.utils.data_pipeline.provides("signal")
    def audio_chunker(file_path, start_offset=0, length=-1):
        
        if "open_files" not in audio_chunker.__dict__:
            audio_chunker.open_files = {}

        # If file is not open, open it and remember the file handle (SoundFile object)
        if file_path not in audio_chunker.open_files:
            audio_chunker.open_files[file_path] = sf.SoundFile(file_path)

        file_hand = audio_chunker.open_files[file_path]

        file_hand.seek(start_offset)
        audio_numpy = file_hand.read(length, dtype=np.float32)
        if file_hand.tell() >= file_hand.frames:
            file_hand.close()
            del audio_chunker.open_files[file_path]

        audio = torch.from_numpy(audio_numpy)

        return audio
```
